controller/api/destination/watcher/pod_watcher.go

```diff
@@ -0,0 +1,543 @@
+		ip                 string
+		port               Port
+		pod                *corev1.Pod
+		listeners          []PodUpdateListener
+		metrics            metrics
+		log                *logging.Entry
+
+		mu sync.RWMutex
+	}
+
+	// PodUpdateListener is the interface subscribers must implement.
+	PodUpdateListener interface {
+		Update(*Address) (bool, error)
+	}
+)
+
+var ipPortVecs = newMetricsVecs("ip_port", []string{"ip", "port"})
+
+func NewPodWatcher(k8sAPI *k8s.API, metadataAPI *k8s.MetadataAPI, log *logging.Entry, defaultOpaquePorts map[uint32]struct{}) (*PodWatcher, error) {
+	pw := &PodWatcher{
+		defaultOpaquePorts: defaultOpaquePorts,
+		k8sAPI:             k8sAPI,
+		metadataAPI:        metadataAPI,
+		publishers:         make(map[IPPort]*podPublisher),
+		log: log.WithFields(logging.Fields{
+			"component": "pod-watcher",
+		}),
+	}
+
+	_, err := k8sAPI.Pod().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    pw.addPod,
+		DeleteFunc: pw.deletePod,
+		UpdateFunc: pw.updatePod,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = k8sAPI.Srv().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    pw.updateServer,
+		DeleteFunc: pw.updateServer,
+		UpdateFunc: func(_, obj interface{}) { pw.updateServer(obj) },
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pw, nil
+}
+
+// Subscribe notifies the listener on changes on any pod backing the passed
+// host/ip:port or changes to its associated opaque protocol config. If service
+// and hostname are empty then ip should be set and vice-versa. If ip is empty,
+// the corresponding ip is found for the given service/hostname, and returned.
+func (pw *PodWatcher) Subscribe(service *ServiceID, hostname, ip string, port Port, listener PodUpdateListener) (string, error) {
+	if hostname != "" {
+		pw.log.Debugf("Establishing watch on %s.%s.%s:%d", hostname, service.Name, service.Namespace, port)
+	} else if service != nil {
+		pw.log.Debugf("Establishing watch on %s.%s:%d", service.Name, service.Namespace, port)
+	} else {
+		pw.log.Debugf("Establishing watch on %s:%d", ip, port)
+	}
+	pp, err := pw.getOrNewPodPublisher(service, hostname, ip, port)
+	if err != nil {
+		return "", err
+	}
+
+	pp.subscribe(listener)
+
+	address, err := pp.createAddress()
+	if err != nil {
+		return "", err
+	}
+
+	sent, err := listener.Update(&address)
+	if err != nil {
+		return "", err
+	}
+	if sent {
+		pp.metrics.incUpdates()
+	}
+
+	return pp.ip, nil
+}
+
+// Subscribe stops notifying the listener on chages on any pod backing the
+// passed ip:port or its associated protocol config
+func (pw *PodWatcher) Unsubscribe(ip string, port Port, listener PodUpdateListener) {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+
+	pw.log.Debugf("Stopping watch on %s:%d", ip, port)
+	pp, ok := pw.getPodPublisher(ip, port)
+	if !ok {
+		pw.log.Errorf("Cannot unsubscribe from unknown ip:port [%s:%d]", ip, port)
+		return
+	}
+	pp.unsubscribe(listener)
```

Comment:
Any risk of deadlocking here?

```
Unsubscribe():
    pw.mu.Lock()
    defer => lock released when stack is unwinded
    unsubscribe():
        pw.mu.Lock() => lock already held
        defer pw.mu.Unlock()
```

However, we try to acquire the lock in `pp.unsubscribe` . They're both write locks. If we are running in the same thread won't this deadlock?

I put together an example that I think is close to this in concept: https://go.dev/play/p/N30YUDzU7Cj

We could have a read lock here and release it after `getPodPublisher`?

AFAIK Go mutexes aren't recursive/re-entrant (I only found this https://groups.google.com/g/golang-nuts/c/vN5ncBdtkcA).

...but maybe I'm also grossly misinterpreting this code 😅

Reply:
Note that those are different locks: one on `PodWatcher`, the other on `podPublisher` :wink: